src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+// Copyright 2020 The FreeAge authors
+// This file is part of FreeAge, licensed under the new BSD license.
+// See the COPYING file in the project root for the license text.
+
+#include "FreeAge/common/free_age.hpp"
+#include "FreeAge/common/unit_types.hpp"
+#include "FreeAge/common/building_types.hpp"
+#include "FreeAge/common/player.hpp"
+#include "FreeAge/common/damage.hpp"
+#include "FreeAge/common/resources.hpp"
+#include "FreeAge/common/object_types.hpp"
+
+// Naming conversions:
+// *Stats: collection of values
+// *Info: the base *Stats, a list of modications on it and the resulting *Stats
+
+// Player has a CivilizationInfo, a vector<UnitTypeInfo>, a vector<BuildingTypeInfo> and a vector<TechnologyInfo>
+// The game data provide the base UnitTypeStats, BuildingTypeStats, TechnologyStats and CivilizationStats
+
+// Note: The *Stats object do NOT contain information about the dependencies and building's interfaces (unit production,
+//       technology research). The consept of the *Stats and *Info is to have the needed information by the game logic
+//       in a ready to use *Stats struct, but also be able to provide to the gui information about the how the values
+//       from the staring base *Stats have been calculated (by keep track of all the modications).
+//       Dependencies and building's interfaces are accessed by different parts of the game logic and also there is no
+//       need to know anything about the previous state.
+
+// General structure
+//
+// Player
+//  \ CivilizationInfo
+//  \ vector<UnitTypeInfo>
+//  \ vector<BuildingTypeInfo>
+//  \ vector<TechnologyInfo>
+//
+// XxInfo
+// \ XxStats baseStats                        (provided by the game data)
+// \ std::vector<Modification> modifications  (starts empty)
+// \ XxStats stats                            (updates every time the modifications change)
+//
+// XxStats
+// \ raw values
+//
+// Modification
+// \ Type        (what to modify)   (TODO: rename to Target?)
+// \ Operation   (how to modify it)
+// \ int value                (with two ints everything is possible)
+// \ int extra
+//
+
+struct UnitTypeInfo {
+
+  UnitTypeInfo(UnitType type, UnitTypeStats baseStats)
+      : type(type),
+        baseStats(baseStats) {
+    stats = baseStats;
+  }
+
+  UnitType type;
+  UnitTypeStats baseStats;   // immutable
```

Comment:
If these are the same for each player, maybe they can be stored only once instead of duplicating them for each player?

Reply:
Responding to this and:
> Do we need to store all of the modifications separately for each unit and building type? Or would it be sufficient to remember once globally for each player which technologies have been researched and which bonuses the player has? That should allow to reconstruct the list of all modifications to a particular stat, if necessary.

I wrote this like this because they are needed by the `AddModification`, else would need to pass the baseStats and a way to search for the modifications, which would probably be ok. 

And also wanted to support a way to change the baseStats, for example by calling the `MakeCurrentStastTheBaseStats` after the basic civ bonuses are applied (which may not even be needed, now Im thinking about it again).

However, I agree that the modifications vector could be refactored by function call when needed when the PlayerResearchedThechnologyTrackerThing is implemented.

Reply:
Plus: my idea was that game logic only knows about the *Stats and not the *Info, so the client could have *Info for the friendly units and just *Stats for the enemy units per player. The server would have the *Info for each player and could send a *Stats form the  to player if:
 - sees for the first time unit type of other player
 - sees unit type of other player and the stats have changed

In the second scenario, if sending the whole type is too big (which it should be an issue because its a rare event), the server could generate a diff of the changes using a list of `Modiifcations` with `Operation::Set` and send those.

Reply:
At least in the original game, players do get to see the base values plus upgrade values for attack/armor/etc. for enemy units. I think that it is common for better players than me to click on enemy units to determine which technologies the enemy has researched, which may give some information about the enemy's strategy. So the clients would need to know the (or assume some) base values for the enemy units as well to display them.

Based on that, I would suggest to make the server directly tell a client which (new) technologies a player has researched if they see any unit of that player that is affected by these technologies. This would both be relatively simple and require very little bandwidth. This is unless there is a concern that this may give away too much information (which is discussed in another thread here).

Reply:
I was under the wrong impression that you could not see the upgrade values of enemy units (I think because of Dota 2???).

So now its much more simpler. Yes, I agree passing around the techs instead of the stats is the way to go.